### PR TITLE
Properly use thin space after math text operator

### PR DESCRIPTION
**Pull request by mpetroff**
## PR Summary

A thin space should follow a math operator except when it is followed by something like a parenthesis or a bracket. This pull request makes said fix and should fix issue #17852.

## PR Checklist

- [x] Has Pytest style unit tests
- [x] Code is [Flake 8](http://flake8.pycqa.org/en/latest/) compliant
- [ ] [Not applicable] New features are documented, with examples if plot related
- [ ] [Not applicable] Documentation is sphinx and numpydoc compliant
- [ ] [Not applicable] Added an entry to doc/users/next_whats_new/ if major new feature (follow instructions in README.rst there)
- [ ] [Not applicable] Documented in doc/api/next_api_changes/* if API changed in a backward-incompatible way

### CODE REVIEW

**Comment by tacaswell** [comment]
Lets give @anntzer a few days and then merge this.

**Comment by anntzer** [comment]
I didn't look super carefully (may be nice to figure out what exactly is TeX's behavior) but I guess this inserts a space in `\sin^n x` or in `\sin\sqrt x` where we may not want one?

**Comment by mpetroff** [comment]
@anntzer You're correct; I forgot about those cases. The space before the subscript / superscript can easily be handled by also excluding `_` and `^`, which is a change I just made.

According to p. 170 of the TeXbook, space is always inserted after an operator except when followed by opening / closing atoms (parentheses, etc.) and punctuation. This means that space should be inserted before `\sqrt`. It also means that the thin space should actually be included when there's a subscript / superscript, but it should be included after the subscript / superscript. I verified both of these with TeX.

Screenshot from 2020-07-11 21-35-22

However, I'm not sure how to put the thin space after the subscript / superscript, since I haven't quite figured out how that code works.

**Comment by anntzer** [comment]
1. I'm very much just armchair-quarterbacking here, but perhaps if the next token is sub/superscript then set a flag saying "after this sub/super (possibly both sub and super, if both are present), check whether a thin space should be added"?
2. I guess the same logic (with or without sub/super handling) is also needed for operatorname (just below)?  (Should `function` just be implemented in terms of `operatorname`, i.e. replace the whole body of `function()` by `hlist = self.operatorname(...); hlist.function_name = toks[0]; return hlist`, and just put the thin space logic only in `operatorname`?)
3. Regardless, this is clearly an improvement, and does not preclude implementing sub/super handling later, but my guess is that at least we could try handling `operatorname` in this PR (i.e. I'm not insisting on 1., but perhaps on 2.).

**Comment by mpetroff** [comment]
I extended the improvement to include `operatorname`. However, the tests are going to fail now because the old rendering of `$\operatorname{cos} x$` was incorrect.

**Comment by anntzer** [comment]
I guitars it would be reasonable to just get rid of the single failing test as it's behavior is effectively being tested by the check_figures_equal test.

**Comment by mpetroff** [comment]
With regard to subscript / superscript, I can easily determine in the `subsuper` parsing function that the `nucleus` is an operator. However, I can't figure out how to determine how many characters make up the full subscript / superscript expression, so I can't check what follows to determine whether or not a thin space should be inserted.

**Comment by mpetroff** [comment]
I updated the commit that dropped the test to also drop the corresponding baseline images.

File: lib/matplotlib/mathtext.py [hunk 1]
```diff
@@ -2873,25 +2873,37 @@ def accent(self, s, loc, toks):
...
+            elif isinstance(c, str):
+                hlist_list.append(Char(c, state))
+            else:
+                hlist_list.append(c)
+        next_char_loc = loc + len(toks[0]) + 1
+        if isinstance(toks[0], ParseResults):
+            next_char_loc += 14  # Skip `operatorname{}`
+        next_char = next((c for c in s[next_char_loc:] if c != ' '), '')
+        delimiters = self._left_delim | self._ambi_delim | self._right_delim
+        delimiters |= set(r"^_".split())
```
Comment (QuLogic) [inline lib/matplotlib/mathtext.py:2900]: Is this right? `.split()` splits on whitespace, and there's none in this literal. Are you sure you don't want?
```suggestion
        delimiters |= {"^", "_"}
```

File: lib/matplotlib/mathtext.py [hunk 2]
```diff
@@ -2873,25 +2873,37 @@ def accent(self, s, loc, toks):
...
                 c.font = 'rm'
                 c._update_metrics()
+                hlist_list.append(c)
+            elif isinstance(c, str):
+                hlist_list.append(Char(c, state))
+            else:
+                hlist_list.append(c)
+        next_char_loc = loc + len(toks[0]) + 1
+        if isinstance(toks[0], ParseResults):
+            next_char_loc += 14  # Skip `operatorname{}`
```
Comment (QuLogic) [inline lib/matplotlib/mathtext.py:2897]: I'd just insert it in there:
```suggestion
            next_char_loc += len('operatorname{}')
```

File: lib/matplotlib/tests/test_mathtext.py [hunk 3]
```diff
@@ -285,6 +289,22 @@ def test_spaces(fig_test, fig_ref):
...
 
+@check_figures_equal(extensions=["png"])
+def test_operator_space(fig_test, fig_ref):
+    fig_test.text(0.1, 0.1, r"$\log 6$")
+    fig_test.text(0.1, 0.2, r"$\log(6)$")
+    fig_test.text(0.1, 0.3, r"$\arcsin 6$")
+    fig_test.text(0.1, 0.4, r"$\arcsin|6|$")
+    fig_test.text(0.1, 0.5, r"$\operatorname{op} 6$")  # GitHub issue #553
+    fig_test.text(0.1, 0.6, r"$\operatorname{op}[6]$")
+    fig_ref.text(0.1, 0.1, r"$\mathrm{log\,}6$")
```
Comment (QuLogic) [inline lib/matplotlib/tests/test_mathtext.py:300]: Add blank line between `fig_test` and `fig_ref` section.

File: lib/matplotlib/mathtext.py (hunk 1 above)
Comment (mpetroff) [reply lib/matplotlib/mathtext.py]: You're correct. I broke it sometime between testing it and committing it, and I neglected to add a test case for it. I've now fixed it and added an appropriate unit test.

**Pull request merged**